gmf-permalink - support map_crosshair & map_tooltip

adube opened the pull request:
This PR introduces the support of the `map_crosshair` and `map_tooltip` parameters for the `gmf-permalink` service.

Tasks to do:

 * [x] `map_crosshair`
 * [x] Define a default style for the map crosshair feature added
 * [x] Be able to define a custom style for the map crosshair feature added
 * [x] review `map_crosshair` code
 * [x] use ngeo.FeatureOverlayMgr instead of creating our own overlay
 * [x] `map_tooltip`
 * [x] use jQuery instead of native JavaScript for DOM manipulation
 * [x] review the current style of the popover
 * [x] enhance the CSS style of the popover
 * [x] use `&times;` instead of `x`
 * [x] review `map_tooltip` code
 * [x] merge into one commit
 * [x] travis passes

Live example:

 * http://adube.github.io/ngeo/gmf-permalink-map-crosshair-tooltip/examples/contribs/gmf/permalink.html

adube commented:
`make gh-pages` doesn't work. I get:

```
$ make gh-pages
mkdir -p .build/examples-hosted/contribs/gmf/
cp -r contribs/gmf/build .build/examples-hosted/contribs/gmf/
rm -rf .build/apidoc
./node_modules/.bin/jsdoc -c jsdoc/config.json --destination .build/apidoc
/opt/ngeo/adube/jsdoc/template/publish.js:446
    jsdocAngularJS.registerAngularJSLink(helper);
                   ^

TypeError: jsdocAngularJS.registerAngularJSLink is not a function
    at Object.exports.publish (/opt/ngeo/adube/jsdoc/template/publish.js:446:20)
    at Object.module.exports.cli.generateDocs (/opt/ngeo/adube/node_modules/jsdoc/cli.js:430:39)
    at Object.module.exports.cli.processParseResults (/opt/ngeo/adube/node_modules/jsdoc/cli.js:383:20)
    at module.exports.cli.main (/opt/ngeo/adube/node_modules/jsdoc/cli.js:227:14)
    at Object.module.exports.cli.runCommand (/opt/ngeo/adube/node_modules/jsdoc/cli.js:180:5)
    at /opt/ngeo/adube/node_modules/jsdoc/jsdoc.js:103:9
    at Object.<anonymous> (/opt/ngeo/adube/node_modules/jsdoc/jsdoc.js:104:3)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:313:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
    at node.js:963:3
make: *** [.build/apidoc] Error 1
```


I already did:

```
make cleanall
make cleanallcache
npm install
make check
```

What am I missing ?

adube commented:
@fgravin I also completed the `map_tooltip`.  I would like to show you the current result, but I still have the issue when doing `make gh-pages`.

sbrunner commented:
What about doing: `make .build/apidoc` ?

adube commented:
>What about doing: make .build/apidoc ?

I don't know what was wrong yesterday, but today after rebasing it worked.  I didn't need to do the above.

I added the example.  @pgiraud would you like to check if the popover styling is to your liking ?  A review would be nice for this.

fgravin commented on contribs/gmf/examples/permalink.html:
```diff
@@ -23,6 +32,26 @@
     <p id="desc">
       This example demonstrates the use of the <code>gmf-permalink</code>
       service, which is injected inside the <code>gmf-map</code> directive.
+      The following links demonstrate the different features:
```
demonstrates

adube replied:
the following link**s** (plural), therefore `demonstrate` without an s.

fgravin commented on contribs/gmf/src/services/permalink.js:
```diff
@@ -181,6 +268,47 @@ gmf.Permalink.prototype.registerMap_ = function(map) {
+
+  // (3) Add map crosshair, if set
+  if (this.getMapCrosshair()) {
+    var source = new ol.source.Vector();
```
Better to use the `ngeo.FeatureOverlayMgr` and put the style in the feature.

adube replied:
ok

fgravin commented on contribs/gmf/src/services/permalink.js:
```diff
@@ -181,6 +268,47 @@ gmf.Permalink.prototype.registerMap_ = function(map) {
+    }
+    goog.asserts.assertArray(tooltipPosition);
+
+    var div = document.createElement('DIV');
```
use jquery instead

fgravin commented:
The approach is globally good. Some minor comments.
Also i don't understand why you don't use JQuery for the DOM manipulation...

adube commented:
>The approach is globally good. Some minor comments.
Also i don't understand why you don't use JQuery for the DOM manipulation...

https://github.com/fgravin/ngeo/blob/guidelines/docs/guidelines.md#usage-of-the-closure-library

It states:

>goog functions and objects should sometimes be preferred.

Closure and Angular are mentioned, but not jQuery.  Also, ol3 is in the process of completely removing all uses of `goog` in their code.  When it's possible to use native JavaScript to do something, why not use it, right ?  That's the idea behind this.

To also add: this is a source file in `ol-ext`, so I wanted it to be as close to what the ol3 code would look like.  Please, tell me if I'm wrong.

To my knowledge, there's no jQuery used in ngeo.  There's some in GMF, though.

But then again I use popover, which depends on bootstrap.js, which depends on jquery...  So yeah, I could use jquery.

Please, let me know what you want me to do:

 * leave this as-is
 * use jquery
 * use closure library

fgravin commented:
Yes you already use Jquery with `$(element).popover()` :)

It's quite differents from other native things. Javascript provides more and more native functions for Array, String or things like that ease developement (no more need to use lib functions for that).

But native DOM manipulation is there from the beginning, and library came after to try to ease that, like JQuery.

please use JQuery.

adube commented:
>please use JQuery.

Gotcha

adube commented:
@fgravin Thanks for the review.  I addressed all required changes.  I also updated the live demo.

Look at the popover.  Do you think its styling is satisfactory ?  If it does, then I'd merge all commits and this would be ready to merge into master.

fgravin commented:
Ok thanks great work.
@pgiraud could you please review the design ?

pgiraud commented:
I only concentrated on the styling. I find the "close button" quite ugly. This shouldn't be a button. What about using the "close" className ?

adube commented:
@pgiraud Makes sense. I'm on it.

adube commented:
@pgiraud Fix made.  How about now ?

@fgravin I merged all commits.  If @pgiraud "rolls and doesn't collect dust" (agrees), then we could merge.

pgiraud commented on src/ol-ext/popover.js:
```diff
@@ -0,0 +1,113 @@
+   */
+  this.closeEl_ = $('<button>', {
+    'class': 'close',
+    'text': 'x'
```
Please use `&times;` instead.

adube replied:
Fixed.  Thanks.

pgiraud commented on src/ol-ext/popover.js:
```diff
@@ -0,0 +1,113 @@
+            '  <h3 class="popover-title"></h3>',
+            '  <div class="popover-content"></div>',
+            '</div>'
+          ].join('')
```
Are you sure you need to provide the whole template?
For the `ngeo-popover` className? Isn't there a less intrusive way to do it?

pgiraud replied:
I don't find any other way. So this looks good to me.

adube replied:
Yeah, I also thought this was unfortunate.

adube commented:
@fgravin This is ready for merge.

fgravin commented on src/ol-ext/popover.js:
```diff
@@ -0,0 +1,113 @@
+  this.clickKey_ = null;
+
+  var originalEl;
+  if (options.element) {
```
`options.element` and `originalEl` could be jQuery elements.

fgravin commented on src/ol-ext/popover.js:
```diff
@@ -0,0 +1,113 @@
+    .append(this.closeEl_)
+    .append(originalEl)[0];
+
+  options.element = $('<div />')[0];
```
actually i guess it's the only one that need to be a native element (for ol3).

fgravin commented on src/ol-ext/popover.js:
```diff
@@ -0,0 +1,113 @@
+   */
+  this.contentEl_ = $('<div/>')
+    .append(this.closeEl_)
+    .append(originalEl)[0];
```
too much useless `[0]` no ?

fgravin commented on contribs/gmf/src/services/permalink.js:
```diff
@@ -181,6 +269,45 @@ gmf.Permalink.prototype.registerMap_ = function(map) {
+    }
+    goog.asserts.assertArray(tooltipPosition);
+
+    var div = $('<div/>')[0];
```
all can be a jQuery type

adube replied:
It can't.  An `ol.Overlay` takes an `Element`, not a `jQuery` as the `element` property.

pgiraud replied:
I think that what @fgravin meant is the following:
```
var div = $('<div>', {
  text: tooltipText,
  'class': 'gmf-permalink-tooltip'
})[0];
```

adube replied:
Fixed.

sbrunner commented:
/!\ Don't merge it this week for the Monday demo :-)

adube commented:
All comments by reviewer have been fixed.  Let's wait after Monday to merge.

sbrunner commented:
For me it can be merged :-)

fredj commented:
LGTM, thanks @adube
